### PR TITLE
[VAULT-2533] Add Redshift support to Provider

### PR DESCRIPTION
**Pull request by vinay-gopalan**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
`resource/database_secret_backend_connection`: Add support for configuring Redshift databases
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-v -run TestAccDatabaseSecretBackendConnection_redshift''
RUN   TestAccDatabaseSecretBackendConnection_redshift
--- PASS: TestAccDatabaseSecretBackendConnection_redshift (2.37s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault
...
```

### CODE REVIEW

File: vault/resource_database_secret_backend_connection.go [hunk 1]
```diff
@@ -388,6 +390,15 @@ func databaseSecretBackendConnectionResource() *schema.Resource {
 				ConflictsWith: util.CalculateConflictsWith(dbBackendOracle, dbBackendTypes),
 			},
 
+			"redshift": {
+				Type:          schema.TypeList,
+				Optional:      true,
+				Description:   "Connection parameters for the redshift-database-plugin plugin.",
+				Elem:          connectionStringResource(&connectionStringConfig{}),
+				MaxItems:      1,
+				ConflictsWith: util.CalculateConflictsWith("redshift", dbBackendTypes),
```
Comment (benashz) [inline vault/resource_database_secret_backend_connection.go:399]: Use the constant here:

```suggestion
				ConflictsWith: util.CalculateConflictsWith(dbBackendRedshift, dbBackendTypes),
```

File: vault/resource_database_secret_backend_connection_test.go [hunk 2]
```diff
@@ -1094,6 +1129,24 @@ resource "vault_database_secret_backend_connection" "test" {
...
+	return fmt.Sprintf(`
+resource "vault_mount" "db" {
+  path = "%s"
+  type = "database"
+}
+resource "vault_database_secret_backend_connection" "test" {
+  backend = "${vault_mount.db.path}"
+  name = "%s"
+  allowed_roles = ["dev", "prod"]
+  root_rotation_statements = ["FOOBAR"]
```
Comment (benashz) [inline vault/resource_database_secret_backend_connection_test.go:1142]: Any specific reason that the statement is in all caps?

File: vault/resource_database_secret_backend_connection_test.go [hunk 3]
```diff
@@ -703,6 +703,41 @@ func TestAccDatabaseSecretBackendConnection_snowflake(t *testing.T) {
 	})
 }
 
+func TestAccDatabaseSecretBackendConnection_redshift(t *testing.T) {
+	MaybeSkipDBTests(t, dbBackendSnowflake)
```
Comment (benashz) [inline vault/resource_database_secret_backend_connection_test.go:707]: Shouldn't that be `dbBackendRedshift` ?

File: vault/resource_database_secret_backend_connection.go (hunk 1 above)
Comment (vinay-gopalan) [reply vault/resource_database_secret_backend_connection.go]: Fixed in 20fcb20

File: vault/resource_database_secret_backend_connection_test.go (hunk 3 above)
Comment (vinay-gopalan) [reply vault/resource_database_secret_backend_connection_test.go]: Fixed in 20fcb20

File: vault/resource_database_secret_backend_connection_test.go (hunk 2 above)
Comment (vinay-gopalan) [reply vault/resource_database_secret_backend_connection_test.go:1142]: No reason, I was just following the convention for all the other DB tests: https://github.com/hashicorp/terraform-provider-vault/blob/main/vault/resource_database_secret_backend_connection_test.go#L735

File: vault/resource_database_secret_backend_connection.go [hunk 4]
```diff
@@ -609,6 +622,8 @@ func getDatabaseAPIData(d *schema.ResourceData) (map[string]interface{}, error)
 		setElasticsearchDatabaseConnectionData(d, "elasticsearch.0.", data)
 	case "snowflake-database-plugin":
 		setSnowflakeDatabaseConnectionData(d, "snowflake.0.", data)
+	case "redshift-database-plugin":
+		setDatabaseConnectionData(d, "redshift.0.", data)
```
Comment (benashz) [inline vault/resource_database_secret_backend_connection.go:626]: It looks like `setDatabaseConnectionData()` does not include support for [username](https://www.vaultproject.io/api-docs/secret/databases/redshift#username) or [password](https://www.vaultproject.io/api-docs/secret/databases/redshift#password)

Seems to be the same issue for [postgresql](https://www.vaultproject.io/api-docs/secret/databases/postgresql) as well, so the omissions might be intentional.

Comment (vinay-gopalan) [reply vault/resource_database_secret_backend_connection.go:626]: Good catch! That definitely looked like a bug. I confirmed that all DBs using `setDatabaseConnectionData` have username and password fields, and I added support for username and password to that function.

File: website/docs/r/database_secret_backend_connection.md [hunk 5]
```diff
@@ -295,6 +295,27 @@ See the [Vault
...
+  the database.
+
+* `max_idle_connections` - (Optional) The maximum number of idle connections to
+  the database.
+
+* `max_connection_lifetime` - (Optional) The maximum amount of time a connection may be reused.
+
+* `username` - (Optional) The root credential username used in the connection URL.
+
+* `password` - (Optional) The root credential password used in the connection URL.
```
Comment (benashz) [inline website/docs/r/database_secret_backend_connection.md:315]: Probably want to revert these fields

File: vault/resource_database_secret_backend_connection.go [hunk 6]
```diff
@@ -907,19 +936,14 @@ func setInfluxDBDatabaseConnectionData(d *schema.ResourceData, prefix string, da
 	}
 }
 
-func setSnowflakeDatabaseConnectionData(d *schema.ResourceData, prefix string, data map[string]interface{}) {
+func setDatabaseConnectionData_userPass(d *schema.ResourceData, prefix string, data map[string]interface{}) {
```
Comment (benashz) [inline vault/resource_database_secret_backend_connection.go:939]: Underscores in function names are not standard in Go

```suggestion
func setDatabaseConnectionDataWithUserPass(d *schema.ResourceData, prefix string, data map[string]interface{}) {
```

Comment (vinay-gopalan) [reply vault/resource_database_secret_backend_connection.go]: Fixed in 8f3da39

File: website/docs/r/database_secret_backend_connection.md (hunk 5 above)
Comment (vinay-gopalan) [reply website/docs/r/database_secret_backend_connection.md:315]: Do we still want to revert these? Now we are explicitly setting these fields in the schema

Comment (benashz) [reply website/docs/r/database_secret_backend_connection.md:315]: No, should be good now that they are part of the schema.

**Review by benashz — APPROVED** [review]
Ship it!

File: vault/resource_database_secret_backend_connection_test.go [hunk 7]
```diff
@@ -703,6 +703,41 @@ func TestAccDatabaseSecretBackendConnection_snowflake(t *testing.T) {
...
+
+	url := os.Getenv("REDSHIFT_URL")
+	if url == "" {
+		t.Skip("REDSHIFT_URL not set")
+	}
+	backend := acctest.RandomWithPrefix("tf-test-db")
+	name := acctest.RandomWithPrefix("db")
+	resource.Test(t, resource.TestCase{
+		Providers:    testProviders,
+		PreCheck:     func() { testAccPreCheck(t) },
```
Comment (benashz) [inline vault/resource_database_secret_backend_connection_test.go:717]: `testAccPreCheck()` was just replaced by `testutil.TestAccPreCheck()` in 5660a057265b1eb2c00c7c2015de01979caa42f6, you'll want to merge/rebase.

**Review by benashz — APPROVED** [review]
This time for sure 👍

File: vault/resource_database_secret_backend_connection_test.go (hunk 7 above)
Comment (vinay-gopalan) [reply vault/resource_database_secret_backend_connection_test.go]: Fixed in 8185ea1

**Review by benashz — APPROVED** [review]

**Pull request merged**